File: mobile/shared/src/commonTest/kotlin/com/bissbilanz/analytics/CaloriePatternsTest.kt
package com.bissbilanz.analytics

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class CaloriePatternsTest {
    @Test
    fun frontLoadingMorningHeavy() {
        val entries =
            listOf(
                Triple("2024-01-01", "2024-01-01T07:00:00Z", 600.0),
                Triple("2024-01-01", "2024-01-01T10:00:00Z", 400.0),
                Triple("2024-01-01", "2024-01-01T19:00:00Z", 200.0),
                Triple("2024-01-02", "2024-01-02T08:00:00Z", 500.0),
                Triple("2024-01-02", "2024-01-02T11:00:00Z", 400.0),
                Triple("2024-01-02", "2024-01-02T20:00:00Z", 100.0),
            )
        val result = computeCalorieFrontLoading(entries)
        assertTrue(result.avgMorningPct > 50.0, "Expected morning-heavy loading, got ${result.avgMorningPct}")
        assertEquals(2, result.daysAbove50Pct)
        assertEquals(2, result.totalDays)
    }

    @Test
    fun frontLoadingEmptyReturnsZero() {
        val result = computeCalorieFrontLoading(emptyList())
        assertEquals(0.0, result.avgMorningPct)
        assertEquals(0, result.daysAbove50Pct)
        assertEquals(0, result.totalDays)
        assertEquals(ConfidenceLevel.INSUFFICIENT, result.confidence)
    }

    @Test
    fun frontLoadingSkipsNullEatenAt() {
        val entries =
            listOf(
                Triple<String, String?, Double>("2024-01-01", null, 500.0),
                Triple<String, String?, Double>("2024-01-02", "2024-01-02T08:00:00Z", 800.0),
            )
        val result = computeCalorieFrontLoading(entries)
        assertEquals(1, result.totalDays)
    }

    @Test
    fun cyclingConsistentPattern() {
        val days =
            (1..14).map { i ->
                Pair("2024-01-${i.toString().padStart(2, '0')}", 2000.0 + i * 10.0)
            }
        val result = computeCalorieCycling(days)
        assertEquals("consistent", result.pattern, "CV=${result.cv} should be consistent (<15)")
    }

    @Test
    fun cyclingHighVariancePattern() {
        val days =
            listOf(
                Pair("2024-01-01", 1000.0),
                Pair("2024-01-02", 3000.0),
                Pair("2024-01-03", 800.0),
                Pair("2024-01-04", 3500.0),
                Pair("2024-01-05", 900.0),
                Pair("2024-01-06", 3200.0),
                Pair("2024-01-07", 1100.0),
            )
        val result = computeCalorieCycling(days)
        assertEquals("high_variance", result.pattern, "Expected high variance, CV=${result.cv}")
        assertTrue(result.cv > 30.0)
    }

    @Test
    fun cyclingModeratePattern() {
        // mean=2000, values spread so CV is in 15-30% range
        // deviations: -500,500,-400,400,-300,300 → mean of 6 values=2000, variance=175000, stddev=418, CV=20.9%
        val days =
            listOf(
                Pair("2024-01-01", 1500.0),
                Pair("2024-01-02", 2500.0),
                Pair("2024-01-03", 1600.0),
                Pair("2024-01-04", 2400.0),
                Pair("2024-01-05", 1700.0),
                Pair("2024-01-06", 2300.0),
            )
        val result = computeCalorieCycling(days)
        assertEquals("moderate", result.pattern, "CV=${result.cv} should be moderate (15-30)")
        assertTrue(result.cv >= 15.0 && result.cv < 30.0)
    }

    @Test
    fun cyclingEmptyReturnsZero() {
        val result = computeCalorieCycling(emptyList())
        assertEquals(0.0, result.mean)
        assertEquals(0.0, result.stddev)
        assertEquals(0.0, result.cv)
        assertEquals("consistent", result.pattern)
        assertEquals(ConfidenceLevel.INSUFFICIENT, result.confidence)
    }

    @Test
    fun frontLoadingSingleDayProducesResult() {
        val entries = listOf(Triple("2024-01-01", "2024-01-01T10:00:00Z", 500.0))
        val result = computeCalorieFrontLoading(entries)
        assertEquals(1, result.totalDays)
        assertEquals(100.0, result.avgMorningPct, 1e-9)
    }

    @Test
    fun frontLoadingAllNullEatenAtReturnsEmpty() {
        val entries =
            listOf(
                Triple("2024-01-01", null as String?, 500.0),
                Triple("2024-01-02", null, 600.0),
            )
        val result = computeCalorieFrontLoading(entries)
        assertEquals(0, result.totalDays)
        assertEquals(0.0, result.avgMorningPct)
    }

    @Test
    fun frontLoadingCustomCutoffHour() {
        val entries =
            listOf(
                Triple("2024-01-01", "2024-01-01T10:00:00Z", 500.0),
                Triple("2024-01-01", "2024-01-01T11:00:00Z", 500.0),
            )
        val result10 = computeCalorieFrontLoading(entries, cutoffHour = 10)
        assertEquals(0.0, result10.avgMorningPct, 1e-9)
        val result12 = computeCalorieFrontLoading(entries, cutoffHour = 12)
        assertEquals(100.0, result12.avgMorningPct, 1e-9)
    }

    @Test
    fun cyclingSingleDayReturnsConsistentWithZeroCV() {
        val result = computeCalorieCycling(listOf(Pair("2024-01-01", 2000.0)))
        assertEquals(1, result.sampleSize)
        assertEquals(0.0, result.cv)
        assertEquals("consistent", result.pattern)
    }

    @Test
    fun cyclingCVAtBoundary15Percent() {
        val result = computeCalorieCycling(listOf(Pair("d1", 1700.0), Pair("d2", 2300.0)))
        assertEquals("moderate", result.pattern)
    }
}

File: tests/analytics/calorie-patterns.test.ts
import { describe, test, expect } from 'vitest';
import { computeCalorieFrontLoading, computeCalorieCycling } from '$lib/analytics/calorie-patterns';

describe('computeCalorieFrontLoading', () => {
	test('all morning eating gives high avgMorningPct', () => {
		const entries = [
			{ date: '2024-01-01', eatenAt: '2024-01-01T08:00:00+00:00', calories: 500 },
			{ date: '2024-01-01', eatenAt: '2024-01-01T10:00:00+00:00', calories: 500 },
			{ date: '2024-01-02', eatenAt: '2024-01-02T09:00:00+00:00', calories: 600 },
			{ date: '2024-01-02', eatenAt: '2024-01-02T11:00:00+00:00', calories: 400 }
		];
		const result = computeCalorieFrontLoading(entries, 14);
		expect(result.avgMorningPct).toBe(100);
		expect(result.daysAbove50Pct).toBe(2);
	});

	test('all evening eating gives low avgMorningPct', () => {
		const entries = [
			{ date: '2024-01-01', eatenAt: '2024-01-01T18:00:00+00:00', calories: 800 },
			{ date: '2024-01-01', eatenAt: '2024-01-01T20:00:00+00:00', calories: 700 }
		];
		const result = computeCalorieFrontLoading(entries, 14);
		expect(result.avgMorningPct).toBe(0);
		expect(result.daysAbove50Pct).toBe(0);
	});

	test('skips entries without eatenAt', () => {
		const entries = [
			{ date: '2024-01-01', eatenAt: null, calories: 800 },
			{ date: '2024-01-01', eatenAt: '2024-01-01T09:00:00+00:00', calories: 500 }
		];
		const result = computeCalorieFrontLoading(entries, 14);
		expect(result.sampleSize).toBe(1);
		expect(result.avgMorningPct).toBe(100);
	});

	test('returns zero for empty input', () => {
		const result = computeCalorieFrontLoading([], 14);
		expect(result.avgMorningPct).toBe(0);
		expect(result.sampleSize).toBe(0);
		expect(result.confidence).toBe('insufficient');
	});
});

describe('computeCalorieCycling', () => {
	test('consistent 2000 kcal gives low CV and consistent pattern', () => {
		const days = Array.from({ length: 14 }, (_, i) => ({
			date: `2024-01-${String(i + 1).padStart(2, '0')}`,
			calories: 2000
		}));
		const result = computeCalorieCycling(days);
		expect(result.cv).toBe(0);
		expect(result.pattern).toBe('consistent');
		expect(result.stddev).toBe(0);
	});

	test('alternating 1500/2500 gives high variance', () => {
		const days = Array.from({ length: 20 }, (_, i) => ({
			date: `2024-01-${String(i + 1).padStart(2, '0')}`,
			calories: i % 2 === 0 ? 1500 : 2500
		}));
		const result = computeCalorieCycling(days);
		expect(result.cv).toBeGreaterThan(15);
		expect(result.mean).toBe(2000);
	});

	test('pattern thresholds', () => {
		const makeUniform = (cal: number, count = 10) =>
			Array.from({ length: count }, (_, i) => ({
				date: `2024-01-${String(i + 1).padStart(2, '0')}`,
				calories: cal
			}));

		expect(computeCalorieCycling(makeUniform(2000)).pattern).toBe('consistent');

		const moderate = Array.from({ length: 10 }, (_, i) => ({
			date: `2024-01-${String(i + 1).padStart(2, '0')}`,
			calories: 2000 + (i % 2 === 0 ? 250 : -250)
		}));
		const modResult = computeCalorieCycling(moderate);
		expect(modResult.cv).toBeGreaterThan(0);
		expect(['moderate', 'consistent']).toContain(modResult.pattern);
	});

	test('highDays and lowDays counts', () => {
		const days = [
			{ date: '2024-01-01', calories: 2000 },
			{ date: '2024-01-02', calories: 2000 },
			{ date: '2024-01-03', calories: 2000 },
			{ date: '2024-01-04', calories: 3000 },
			{ date: '2024-01-05', calories: 1000 }
		];
		const result = computeCalorieCycling(days);
		expect(result.highDays).toBe(1);
		expect(result.lowDays).toBe(1);
	});

	test('returns zero values for empty input', () => {
		const result = computeCalorieCycling([]);
		expect(result.sampleSize).toBe(0);
		expect(result.confidence).toBe('insufficient');
	});
});
